pkg/controller/pinned-image-set/pinned_image_set_controller.go

```diff
@@ -0,0 +1,698 @@
+func (ctrl *Controller) ensurePinnedImageSet(imageSet *mcfgv1.PinnedImageSet) error {
+	// Check for Deleted PinnedImageSet and optionally delete finalizers
+	if imageSet.DeletionTimestamp != nil {
+		if len(imageSet.GetFinalizers()) > 0 {
+			klog.V(2).Infof("sync() deleting finalizers for PinnedImageSet: %s\n", imageSet.Name)
+			return ctrl.cascadeDelete(imageSet)
+		}
+		return nil
+	}
+
+	// get pools that match node selector
+	mcpPools, err := ctrlcommon.GetPoolsForPinnedImageSet(ctrl.mcpLister, imageSet)
+	if err != nil {
+		return err
+	}
+	if len(mcpPools) == 0 {
+		return fmt.Errorf("%w : %s", ErrNoMatchingPool, imageSet.Name)
+	}
+
+	for _, pool := range mcpPools {
+		managedKey, err := getMachineConfigKey(pool, ctrl.mcfgClient, imageSet)
+		if err != nil {
+			return err
+		}
+
+		mc, err := ctrl.mcfgClient.MachineconfigurationV1().MachineConfigs().Get(context.TODO(), managedKey, metav1.GetOptions{})
+		// check if not found error
+		isNotFound := apierrors.IsNotFound(err)
+		if err != nil && !isNotFound {
+			return err
+		}
+		// If we have seen this generation and the sync didn't fail, then skip
+		if !isNotFound && imageSet.Status.ObservedGeneration >= imageSet.Generation {
+			// But we still need to compare the generated controller version because during an upgrade we need a new one
+			mcCtrlVersion := mc.Annotations[ctrlcommon.GeneratedByControllerVersionAnnotationKey]
+			if mcCtrlVersion == version.Hash {
+				return nil
+			}
+		}
+
+		pinnedImages := imageSet.Spec.PinnedImages
+		if len(pinnedImages) == 0 {
+			return fmt.Errorf("no pinned images for PinnedImageSet %s", imageSet.Name)
+		}
+
+		tomlConf := tomlConfigCRIOPinnedImages{}
+		tomlConf.Crio.Image.PinnedImages = pinnedImages
+		configFile, err := createTOMLgeneratedConfigFile(CRIODropInFilePathPinnedImages, tomlConf)
+		if err != nil {
+			return err
+		}
+
+		if isNotFound {
+			tempIgnCfg := ctrlcommon.NewIgnConfig()
+			mc, err = ctrlcommon.MachineConfigFromIgnConfig(pool.Name, managedKey, tempIgnCfg)
+			if err != nil {
+				return err
+			}
+		}
+
+		imageSetRuntimeConfigIgn := createNewIgnition(configFile)
+		rawCtrRuntimeConfigIgn, err := json.Marshal(imageSetRuntimeConfigIgn)
+		if err != nil {
+			return fmt.Errorf("error marshalling PinnedImageSet Ignition: %w", err)
+		}
+		mc.Spec.Config.Raw = rawCtrRuntimeConfigIgn
+
+		mc.SetAnnotations(map[string]string{
+			ctrlcommon.GeneratedByControllerVersionAnnotationKey: version.Hash,
+		})
+		oref := metav1.NewControllerRef(imageSet, controllerKind)
+		mc.SetOwnerReferences([]metav1.OwnerReference{*oref})
+
+		// Create or Update, on conflict retry
+		if err := retry.RetryOnConflict(ctrl.waitBackOff, func() error {
+			var err error
+			if isNotFound {
+				_, err = ctrl.mcfgClient.MachineconfigurationV1().MachineConfigs().Create(context.TODO(), mc, metav1.CreateOptions{})
+			} else {
+				_, err = ctrl.mcfgClient.MachineconfigurationV1().MachineConfigs().Update(context.TODO(), mc, metav1.UpdateOptions{})
```

Comment:
the preferred mechanism is `patch` I believe. You can look around at how to make a `sonmergepatch.CreateThreeWayJSONMergePatch(curJSON, modJSON, curJSON)` and then pass this output to `.Patch` rather than `.Update`.

There are some scenarios where you want to use update but I am forgetting if this falls into those.

Reply:
Sounds good I will dig into it to ensure correctness this is copy pasta from an existing controller.